// agent-ai-faker/src/helpers/ai/sevices/generator/DataGenerator.ts
import cohere from '../generator/cohereClient';
import { UserData, GenerateDataOptions } from '../generator/dataTypes';

export async function generateFakeData(requirement: string, options: GenerateDataOptions): Promise<string> {
  const prompt = `
  Generate fake data based on the following business requirement:
  "${requirement}"

  Return the data in ${options.format} format. Include the following fields for each user:
  - userId
  - firstName
  - lastName
  - email
  - subscriptionStatus
  - lastEngagementDate

  Ensure the data is structured and does not include explanations or additional text.
  `;

  try {
    const response = await cohere.chat({
      message: prompt, // Use 'message' instead of 'prompt' for chat API
      model: 'command-nightly',
      maxTokens: 1000,
      temperature: 0.3,
    });

    // Extract the generated text from the chat response
    const generatedData = response.text.trim();
    return options.format === 'json' ? generatedData : convertJsonToCsv(generatedData);
  } catch (error) {
    console.error('Error generating fake data:', error);
    throw error;
  }
}

function convertJsonToCsv(jsonData: string): string {
  const data = JSON.parse(jsonData) as UserData[];
  const headers = Object.keys(data[0]);
  const csvRows = [
    headers.join(','), // Header row
    ...data.map(row => headers.map(header => String(row[header as keyof UserData])).join(','))
  ];
  return csvRows.join('\n');
}
