// agent-ai-faker/src/helpers/ai/sevices/FakeDataOptions.ts
import cohere from '../sevices/generator/cohereClient';

interface FakeDataOptions {
    purpose: string; // Purpose of the fake data (e.g., testing, training, demo)
    format: string;  // Desired output format (e.g., JSON, CSV, plain text)
    fields: string[]; // List of fields/attributes (e.g., name, age, address)
    volume: number;  // Number of records needed
    constraints: string; // Specific rules or patterns for the data
}

export async function generateFakeData(options: FakeDataOptions): Promise<string> {
    const { purpose, format, fields, volume, constraints } = options;
    // Construct prompt for Cohere API
    const prompt = `Generate fake data for ${purpose} in ${format} format with fields: ${fields.join(", ")}. Volume: ${volume} records. Constraints: ${constraints}.`;

    try {
        const response = await cohere.generate({
            model: 'command',
            prompt,
            maxTokens: 500,
            temperature: 0.7,
        });

        return response.generations[0].text;
    } catch (error) {
        console.error("Error generating fake data:", error);
        throw error;
    }
}
